### PR TITLE
KAFKA-15106 fix AbstractStickyAssignor isBalanced predict

### PR DESCRIPTION
**Pull request by flashmouse**
https://issues.apache.org/jira/browse/KAFKA-15106

in 3.5.0 AbstractStickyAssignor may run useless loop in ``performReassignments `` because ``isBalanced`` have a trivial mistake, and result in rebalance timeout in some situation. 

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status

### CODE REVIEW

**Comment by flashmouse** [comment]
@rajinisivaram  @dajac  plz have a look, thx!

**Comment by divijvaidya** [comment]
Thank you for your first change to Apache Kafka @flashmouse! In the JIRA, you mention that this could be reproduced using a unit test. Can you please add the unit test here which would fail prior to this change and succeed after it?

**Comment by flashmouse** [comment]
@divijvaidya  thank you for reply!
the unit test ``org.apache.kafka.clients.consumer.internals.AbstractStickyAssignorTest#testLargeAssignmentAndGroupWithNonEqualSubscription`` could use much more time than configured Timeout(90) and this could reproduce easy when ``partitionCount`` and ``consumerCount`` are large enough and not equal(ex. ``partitionCount``= 200,  ``consumerCount``= 20), in this situation ``isBalanced`` and ``performReassignments`` may run whole loop body.

because both ``performReassignments`` and ``isBalanced`` in ``AbstractStickyAssignor`` are not so efficient(``performReassignments`` is worse), ``isBalanced`` could abate ``performReassignments`` call times, so fix its logic could speed up this unit test when parameters are small enough.

in my test,  after this fix, ``testLargeAssignmentAndGroupWithNonEqualSubscription`` could pass when ``partitionCount``= 200, ``consumerCount`` = 20, but is still slower than the combination ``partitionCount``= 2000, ``consumerCount`` = 2000 because in this test when the 2 parameters are equal, ``isBalanced`` satisfy the prediction ``min >= max - 1`` thus could return true immediately.

may be I need add a unit test code with a new commit?

**Comment by kirktrue** [comment]
@ableegoldman you're pretty familiar with this code, IIUC. If so, could you take a look at this PR? Thanks!

**Comment by kirktrue** [comment]
@flashmouse I'm a little slow on the uptake, so I'm trying to come up with a scenario. Let's say the following is true:

* Topic `foo` has three partitions
* The consumer group has two consumers, `consumer` and `otherConsumer`

In the first case, let's say that `consumer` has one partition and `otherConsumer` has two. In that case it's as balanced as possible, right? But the code in `isBalanced` as written disagrees:

```java
if (consumerPartitionCount < otherConsumerPartitionCount) {
    return false;
}
```

Because `1` is less than `2` 😄 But it's going to run into problems because by returning `false` from `isBalanced`, I assume we then trigger a rebalance, which—at best— will cause the ownership of that partition to move from `otherConsumer` to `consumer`. When the next balance check for is run, we're in the same position 🤷‍♂️ 

But with your fix:

```java
if (consumerPartitionCount + 1 < otherConsumerPartitionCount) {
    return false;
}
```

It's effectively stating that the `consumer` has to have _at least_ two partitions difference, correct? Because `1 + 1` is _not less_ then `2`.

Do I have the above (mostly) correct?

Thanks!

**Comment by flashmouse** [comment]
@kirktrue thank you for reply!

In your case, although ``isBalanced``  return false, ``performReassignments`` still won't reassign any partition, because it would only do reassign when ``currentAssignment.get(consumer).size() > currentAssignment.get(otherConsumer).size() + 1``, it will just loop all partitions in reassignable list and find no one match it so ``modified`` keep ``false`` then return done.

I think the problem is ``isBalanced`` and ``performReassignments`` should have the same predict logic when they judge whether should do reassign. for now  ``performReassignments`` itself think should do reassign only when 2 consumers that their partition num gap should have at least 2 but  ``isBalanced`` is 1. 

I mean with such situation, I think if we not modify ``isBalanced``,  the reassignment logic is still right, but may very slow.    ``performReassignments`` will do quite a lot unnecessary work, as I said above, it could done finally but cost a long time. you can try unit test ``testLargeAssignmentAndGroupWithNonEqualSubscription`` without this pr change, if set ``partitionCount``= 20, ``consumerCount`` = 2, it could done with success, but if set ``partitionCount``= 200, ``consumerCount`` = 20, it may run very long time. I have run the test and no result after 14min, and it could speed up to 47sec after I apply my change.

**Review by rreddy-22 — COMMENTED** [review]
Thanks for the catch! I also observed that the actual reassignments don't take place inside perform reassignments so ideally it should eventually end but only after a long period of time. I ran it for 6hrs and it finally failed with the timeout error. The change would make it faster to determine whether the assignment is balanced or not so I think we should change this irrespective. I think originally the definition of a balanced assignment was that `the numbers of topic partitions assigned to consumers differ by at most one` but this isn't true for assignments with different subscriptions so I agree we should change it.

**Comment by flashmouse** [comment]
@rreddy-22  that's what I mean, thank you, could any one merge this if possible or help review to verify this pr is useful?

we are now using CooperativeSticky strategy in kafka-client3.5 but meet serious problem that it may trigger rebalance with colossal times(complete one rebalance about 3sec, then keep trigger new balance within more than 10minutes, that's say, at least rebalance 200 times). for now we cannot give much more useful information and we want to verify whether that is caused by this problem after this pr is checked.

**Comment by dajac** [comment]
@flashmouse Thanks for the PR. It seems that this condition has been around for a while now. Did you observe a similar issue with version prior to 3.5? I wonder what if changes introduced in 3.5 for the rack aware assignment made this worse than it used to be. I agree that we need to get to the bottom of this.

**Comment by flashmouse** [comment]
@dajac 
we haven't seen such many rebalance times before, we before use CooperativeSticky with kafka-client 3.0.0, rebalance times increase very slight.

but we are not use kafka-client 3.5 directly because CooperativeStickyAssignor  in 3.5 need upgrad server to 3.5 too if need rack-aware assign logic and we couldn't upgrade server right now so we move AbstractStickyAssignor and CooperativeStickyAssignor logic to our application and move the rackInfo to userData. although we have passd all unit test in kafka-client after do this change, we still doubt whether this change caused rebalance time increase so we're keeping check.

**Comment by rreddy-22** [comment]
> @flashmouse Thanks for the PR. It seems that this condition has been around for a while now. Did you observe a similar issue with version prior to 3.5? I wonder what if changes introduced in 3.5 for the rack aware assignment made this worse than it used to be. I agree that we need to get to the bottom of this.

The rack awareness changes introduce a whole iteration through all the potential consumers to check if any consumer has a matching rack and in the worst case scenario where no consumer with a matching rack is found, the whole iteration happens all over again. Previously we would only iterate through the list once, so potentially the time could've doubled now.
https://github.com/apache/kafka/blob/3.5/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java#L1389-L1400

**Comment by flashmouse** [comment]
I found another one bug.

In ``org.apache.kafka.clients.consumer.internals.AbstractStickyAssignor#allSubscriptionsEqual``,  it would check whether one partition is assigned by multiple consumers in different generation, the logic has 3 if-else body, the last one condition ``memberGeneration < otherMemberGeneration`` should add this line below:
```
allPreviousPartitionsToOwner.put(tp, otherConsumer);
```
without this line may happen error in such scenario below:
say we have 3 consumers, they own partition1 respectively in generation 1, 2, 3, and then the check sequence is gen3, gen1, gen2, then gen3 and gen2 are all added to ``consumerToOwnedPartitions`` so this partition added twice.

this may happen if one partition is owned by at least 3 generation consumer.

@rreddy-22  @kirktrue could u take a look? thx!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (rreddy-22) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: So in line 161 we directly added the current consumer to the map allPreviousPartitionsToOwner. And if we choose to keep that line we should remove the put in the first two conditions in lines 175 and 180 and we can keep the change you put. 
However, I feel like this would be redundant since we're replacing the otherMember and adding it back if the conditions aren't favorable. My suggestion would be to replace line 161 with get instead of put and then keep the rest of the code as it is.

**Comment by philipnee** [comment]
Hey @flashmouse - could you explain what do you mean by `useless loop`?

**Comment by flashmouse** [comment]
> Hey @flashmouse - could you explain what do you mean by `useless loop`?

run while body in ``performReassignments`` but couldn't modify assignment

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (flashmouse) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: updated, plz check again

**Comment by philipnee** [comment]
@flashmouse thansk for the PR - I was the one who wrote the if-else if -else block.  I'm a bit surprised that I actually made this mistake, so I held it off for a few days just to make sure.  As you have already provided a case that breaks that existing logic, could you implement this as part of the test? Also, could you provide a test case that fails `if (consumerPartitionCount < otherConsumerPartitionCount) {` but succeeds in `if (consumerPartitionCount + 1 < otherConsumerPartitionCount) {` 

This would help massively for the reviews and documentation.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Does moving this line after the log.warn do anything?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (rreddy-22) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java:163]: We'd Have to add a put here too

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (rreddy-22) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Agreed we should move it back to the original position

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (flashmouse) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java:163]: fixed, plz take a look

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (flashmouse) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: done

**Comment by flashmouse** [comment]
> @flashmouse thansk for the PR - I was the one who wrote the if-else if -else block. I'm a bit surprised that I actually made this mistake, so I held it off for a few days just to make sure. As you have already provided a case that breaks that existing logic, could you implement this as part of the test? Also, could you provide a test case that fails `if (consumerPartitionCount < otherConsumerPartitionCount) {` but succeeds in `if (consumerPartitionCount + 1 < otherConsumerPartitionCount) {`
> 
> This would help massively for the reviews and documentation.

hi @philipnee  thx for reply!

In ``org.apache.kafka.clients.consumer.internals.AbstractStickyAssignorTest``,I add: 

1. ``testAssignmentAndGroupWithNonEqualSubscriptionNotTimeout``, this test may run very long time without fix.
2. ``testSubscriptionNotEqualAndAssignSamePartitionWith3Generation``, this test should fail without fix.

and please help review #13965 if possible, it mentions another serious bugs in ``org.apache.kafka.clients.consumer.internals.AbstractStickyAssignor.ConstrainedAssignmentBuilder#ConstrainedAssignmentBuilder`` and may cause rebalance stuck forever.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java
Comment (rreddy-22) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: Could you please format it in multiple lines like this, it increases readability

Comment (rreddy-22) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: subscriptions.put(
            consumer1,
            buildSubscriptionV2Above(topics(topic1, topic2),
            partitions(tp(topic1, 0)), generationId, 0)
        );

Comment (flashmouse) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: done, but seems this is enough?
```
subscriptions.put(consumer1, buildSubscriptionV2Above(topics(topic),
                    partitions(tp(topic, 0), tp(topic, 2)), ints[0], 0));
```

**Comment by rreddy-22** [comment]
Looks good to me! Thanks @flashmouse for the changes and replies! @dajac is a committer so he'll give the final approval!

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: nit: can the style be consistent with the .put above? i.e.: either (consumer or (

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java:773]: Can we directly test the `GeneralAssignmentBuilder.build()` ? It seems like that's where isBalanced is triggered.  We could test it at the top level, but as the bug occurs in isBalanced() which is only triggered in the GeneralAssignmentBuilder, it would be more direct, when failing, for people to debug.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java:777]: Thanks for the test! I wrote a similar test, `testEnsurePartitionsAssignedToHighestGeneration` but just the generation order was 0, 1, 2 instead of, per your suggestions, 0, 2, 1.  Do you think we could parametrize that test or reuse the same test style, to make things more consistent, but change the assignment order?  Also, is the title accurate? This happens in `allSubscriptionsEqual`, so it doesn't matter if the subscriptions are equal or not right? The goal of this is to ensure consumer with the highest generation gets the right assignment, regardless of the generation order.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java:728]: Just an opinion here: Other tests also use timeout annotation, but it might be better to use `TestUtils.waitForCondition(()` from other tests because it directly tests the timeout of the `assignPartitions` invocation.  To me, it is a bit more precise about where the tests are failing.  Also, any specific reason we use 90?  I assume the unit is second? I see there's a variation of timeout we are using, from 30->90 😅.

**Review by philipnee — COMMENTED** [review]
Thanks for the PR updates - I left some comments. Let me know if they make sense.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java
Comment (flashmouse) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java:773]: for now ``GeneralAssignmentBuilder`` is the inner private class of ``AbstractStickyAssignor``, test it directly need modify it to protected access level.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java
Comment (flashmouse) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java:777]: this method is to test ``allSubscriptionsEqual`` so its name is a mistake, thx!

but this bug in ``allSubscriptionsEqual`` not lead to the specified partition that previously assigned to at least 3 generation consumers cannot be owned by the highest generation, but would lead this partition assign to more than one  consumers so  add to ``assignedPartitions`` more than once and also the function ``getUnassignedPartitions`` will get more partitions than actuality, as a result, many partitions already assigned to consumers will be added to ``unassignedPartitions``, all of these partitions would be assigned twice.

I couldn't understand how to reuse such test style, could you provide any example?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java
Comment (flashmouse) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java:728]: Oh, this 90 should be sec, I set 90 sec because in my m1 air this test may run about 25 sec or more time.

I tried ``TestUtils.waitForCondition`` but seems it is not effective and couldn't throw Exception when execution use time more than expected, I'm not familiar with this and couldn't find the reason.

**Comment by flashmouse** [comment]
@philipnee I do some modification follow by your review, but still some incomprehension left, plz check my comment below your review, thx

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java
Comment (flashmouse) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: done

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java:728]: Ah no problem.

25s is surprisingly long!

**Comment by philipnee** [comment]
Thanks, kindly ask @dajac to review this PR.

**Review by philipnee — APPROVED** [review]

**Comment by philipnee** [comment]
@flashmouse - There are 108 test failures. Some of those are "Existing Failures".  Could you look into it? I think it might not be related to this PR because @kirktrue has had a similar issue recently.

**Comment by flashmouse** [comment]
It seems the tests I commit in this pr (testAssignmentAndGroupWithNonEqualSubscriptionNotTimeout) have failure and reason are all 90 second timeout.
emmm, should I make ``partitionCount, consumerCount`` smaller?

**Comment by philipnee** [comment]
Hmm - you can try.  But I've had similar, if not the same, failures and I haven't had a chance to examine the cause.

**Comment by ableegoldman** [comment]
@flashmouse you can decrease the consumer count. For one thing, this covers the edge case where the consumers do not all have the same topic subscription, which seems to be quite rare. For another, 2,000 consumers is fairly large to begin with. And especially large for a consumer group where the consumers are all given different topic subscriptions.

Honestly it's probably more important to test with a higher partition count than a large group of consumers like this. Are you able to get it passing within 90s with 500 consumers and 5,000 partitions? That feels like a more realistic large-scale test case. If not, fix it with just 500 consumers and maybe run the test with 4,000, then 3,000, then 2,000 partitions, etc, until you can get it to pass.

If it's still not passing with 500 consumers and 2000 partitions, that's a bit troubling. But we can cross that bridge if it comes up

**Comment by flashmouse** [comment]
thx @ableegoldman 
I reduce partitions num and increase consumers num, I think all tests influenced by this pr should be PASSED now.

**Comment by guozhangwang** [comment]
@flashmouse Thanks! I echo other's thoughts that in a unit test, we could refrain ourselves from trying very large scale when trying to validate a specific logic or expose a bug. Just giving a data point here, in StreamsAssignmentScaleTest which is an integration test, given the goal is actually to validate that the assignor is efficient enough such that it can complete the assignment task within a certain time, our scale is around 1000 - 6000 partitions with 2 clients only.

I'm actually wondering if this bug could be exposed with a even small scale, like `3 consumers and 4 partitions` at all?

**Comment by flashmouse** [comment]
> @flashmouse Thanks! I echo other's thoughts that in a unit test, we could refrain ourselves from trying very large scale when trying to validate a specific logic or expose a bug. Just giving a data point here, in StreamsAssignmentScaleTest which is an integration test, given the goal is actually to validate that the assignor is efficient enough such that it can complete the assignment task within a certain time, our scale is around 1000 - 6000 partitions with 2 clients only.
> 
> I'm actually wondering if this bug could be exposed with a even small scale, like `3 consumers and 4 partitions` at all?

@guozhangwang  thx for reply!

this bug couldn't reproduce in a very small scale, it's actually not a bug, but a trivial mistake would make rebalance logic run very slow. if test scale are too small, it couldn't prove this.

I think the root cause is ``performReassignments`` in ``GeneralAssignmentBuilder`` is a O(partSize * consumerNum) functions, it is very inefficient.

**Comment by guozhangwang** [comment]
> I think the root cause is performReassignments in GeneralAssignmentBuilder is a O(partSize * consumerNum) functions, it is very inefficient.

Ack, and with your changes, it's now constant and not linear to either num.partitions or num.consumers right?

**Comment by flashmouse** [comment]
> > I think the root cause is performReassignments in GeneralAssignmentBuilder is a O(partSize * consumerNum) functions, it is very inefficient.
> 
> Ack, and with your changes, it's now constant and not linear to either num.partitions or num.consumers right?

I hadn't changed the logic, it is still a slow function. whether execute ``performReassignments`` is checked by ``isBalanced``, but it is not effective previously and this pr just fix this.

**Comment by guozhangwang** [comment]
> I hadn't changed the logic, it is still a slow function. whether execute performReassignments is checked by isBalanced, but it is not effective previously and this pr just fix this.

I'm a bit confused now.. my understanding is that with this line of change https://github.com/apache/kafka/pull/13920/files#diff-2814643946ccefef81f480e1280e9b3e3397f4673166ca1ac47db7e3cb6fb760L1175, when num.partitions and num.consumers are not the same, we would no longer run `isBalanced | performReassignments` once per each combo, and hence it's no longer linear to num.partitions * num.consumers. Of course it would not return immediately (that's only the case when num.partitions and num.consumers happen to be equal), but it would still be much better than the old logic, right?

**Comment by flashmouse** [comment]
> > I hadn't changed the logic, it is still a slow function. whether execute performReassignments is checked by isBalanced, but it is not effective previously and this pr just fix this.
> 
> I'm a bit confused now.. my understanding is that with this line of change https://github.com/apache/kafka/pull/13920/files#diff-2814643946ccefef81f480e1280e9b3e3397f4673166ca1ac47db7e3cb6fb760L1175, when num.partitions and num.consumers are not the same, we would no longer run `isBalanced | performReassignments` once per each combo, and hence it's no longer linear to num.partitions * num.consumers. Of course it would not return immediately (that's only the case when num.partitions and num.consumers happen to be equal), but it would still be much better than the old logic, right?

ah this is right.

**Comment by guozhangwang** [comment]
LGTM. Merged to trunk. Thanks @flashmouse !

**Comment by guozhangwang** [comment]
Please also update the JIRA ticket status with fix version fields.

**Comment by flashmouse** [comment]
thx, already update jira status, and plz help review #13965 if could, thx.
@guozhangwang @philipnee

**Comment by showuon** [comment]
Backported to 3.5 branch.

**Pull request merged**
